src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,580 @@
+        } catch (Exception ex) {
+            return new InvocationBindingFailureMessage(invocationId, target, ex);
+        }
+        
+        Collection<String> streams = readStreamIds(unpacker);
+        
+        return new StreamInvocationMessage(headers, invocationId, target, arguments, streams);
+    }
+    
+    private HubMessage createCancelInvocationMessage(MessageUnpacker unpacker) throws IOException {
+        Map<String, String> headers = readHeaders(unpacker);
+        String invocationId = unpacker.unpackString();
+        
+        return new CancelInvocationMessage(headers, invocationId);
+    }
+    
+    private HubMessage createCloseMessage(MessageUnpacker unpacker, int itemCount) throws IOException {
+        String error = unpacker.unpackString();
+        boolean allowReconnect = false;
+        
+        if (itemCount > 2) {
+            allowReconnect = unpacker.unpackBoolean();
+        }
+        
+        return new CloseMessage(error, allowReconnect);
+    }
+    
+    private String writeInvocationMessage(InvocationMessage message) throws IOException {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        MessagePacker packer = MessagePack.newDefaultPacker(out);
+        
+        packer.packArrayHeader(6);
+        packer.packInt(message.getMessageType().value);
+        
+        writeHeaders(message.getHeaders(), packer);
+        
+        String invocationId = message.getInvocationId();
+        if (invocationId != null && !invocationId.isEmpty()) {
+            packer.packString(invocationId);
+        } else {
+            packer.packNil();
+        }
+        
+        packer.packString(message.getTarget());
+        
+        Object[] arguments = message.getArguments();
+        packer.packArrayHeader(arguments.length);
+        
+        for (Object o: arguments) {
+            writeValue(o, packer);
+        }
+        
+        writeStreamIds(message.getStreamIds(), packer);
+        
+        packer.flush();
+        String content = out.toString(StandardCharsets.ISO_8859_1);
+        out.close();
+        return content;
+    }
+    
+    private String writeStreamItemMessage(StreamItem message) throws IOException {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        MessagePacker packer = MessagePack.newDefaultPacker(out);
+        
+        packer.packArrayHeader(4);
+        packer.packInt(message.getMessageType().value);
+        
+        writeHeaders(message.getHeaders(), packer);
+        
+        packer.packString(message.getInvocationId());
+        
+        writeValue(message.getItem(), packer);
+        
+        packer.flush();
+        String content = out.toString(StandardCharsets.ISO_8859_1);
+        out.close();
+        return content;
+    }
+    
+    private String writeCompletionMessage(CompletionMessage message) throws IOException {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
```

Comment:
This look inefficient. Do we need to design for writing to something instead of returning strings everywhere?

Reply:
The interface returns/reads Strings, but @BrennanConroy and I were considering changing it to use byte[]. I think I could also move to using [MessageBufferPacker](https://www.javadoc.io/static/org.msgpack/msgpack-core/0.8.20/org/msgpack/core/MessageBufferPacker.html), which should be more efficient than using MessagePacker w/ a ByteArrayOutputStream

Reply:
Just updated to use MessageBufferPacker: https://github.com/dotnet/aspnetcore/pull/23532/commits/bb60bc4ec76a297f07cd826037bf470c5a0429e8